File: crates/finsight-app/src/commands/copilot.rs
use crate::error::AppResult;
use crate::AppState;

// Types live in finsight-api now; re-exported so existing imports of
// `finsight_app::commands::copilot::*` (lib.rs, tests) keep resolving.
pub use finsight_api::commands::copilot::{ExecutionItemResult, ExecutionSummary};

use finsight_core::models::{AgentActionBundle, AgentExecutionEntry, AgentSession};

#[tauri::command]
#[specta::specta]
pub async fn list_agent_sessions(
    state: tauri::State<'_, AppState>,
) -> AppResult<Vec<AgentSession>> {
    finsight_api::commands::copilot::list_agent_sessions(&state.api).await
}

#[tauri::command]
#[specta::specta]
pub async fn create_agent_session(
    state: tauri::State<'_, AppState>,
    title: String,
    task_type: String,
) -> AppResult<AgentSession> {
    finsight_api::commands::copilot::create_agent_session(&state.api, title, task_type).await
}

#[tauri::command]
#[specta::specta]
pub async fn close_agent_session(state: tauri::State<'_, AppState>, id: String) -> AppResult<()> {
    finsight_api::commands::copilot::close_agent_session(&state.api, id).await
}

#[tauri::command]
#[specta::specta]
pub async fn list_action_bundles(
    state: tauri::State<'_, AppState>,
    status_filter: Option<String>,
    session_id: Option<String>,
    limit: Option<u32>,
) -> AppResult<Vec<AgentActionBundle>> {
    finsight_api::commands::copilot::list_action_bundles(
        &state.api,
        status_filter,
        session_id,
        limit,
    )
    .await
}

#[tauri::command]
#[specta::specta]
pub async fn get_action_bundle(
    state: tauri::State<'_, AppState>,
    id: String,
) -> AppResult<Option<AgentActionBundle>> {
    finsight_api::commands::copilot::get_action_bundle(&state.api, id).await
}

#[tauri::command]
#[specta::specta]
pub async fn approve_action_item(
    state: tauri::State<'_, AppState>,
    item_id: String,
) -> AppResult<()> {
    finsight_api::commands::copilot::approve_action_item(&state.api, item_id).await
}

#[tauri::command]
#[specta::specta]
pub async fn reject_action_item(
    state: tauri::State<'_, AppState>,
    item_id: String,
) -> AppResult<()> {
    finsight_api::commands::copilot::reject_action_item(&state.api, item_id).await
}

#[tauri::command]
#[specta::specta]
pub async fn list_execution_log(
    state: tauri::State<'_, AppState>,
    bundle_id: String,
) -> AppResult<Vec<AgentExecutionEntry>> {
    finsight_api::commands::copilot::list_execution_log(&state.api, bundle_id).await
}

#[tauri::command]
#[specta::specta]
pub async fn execute_action_bundle(
    state: tauri::State<'_, AppState>,
    bundle_id: String,
) -> AppResult<ExecutionSummary> {
    finsight_api::commands::copilot::execute_action_bundle(&state.api, bundle_id).await
}

File: ui/src/api/hooks/copilot.ts
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";
import { api, type AgentActionBundle, type AgentExecutionEntry, type AgentSession } from "../openapiClient";
import { unwrap } from "../openapiClient";
import { invalidateDomains } from "../invalidation";
import { actionBundleKeys } from "./_factory";
// Re-export for consumers that historically imported from this module.
export { actionBundleKeys };

export function useAgentSessions() {
  return useQuery<AgentSession[]>({
    queryKey: ["agent-sessions"],
    queryFn: async () => {
      return unwrap(api.listAgentSessions());
    },
  });
}

export function useCreateAgentSession() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async ({ title, taskType }: { title: string; taskType: string }) => {
      return unwrap(api.createAgentSession(title, taskType));
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ["agent-sessions"] });
    },
  });
}

export function useCloseAgentSession() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async (id: string) => {
      await unwrap(api.closeAgentSession(id));
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ["agent-sessions"] });
    },
  });
}

export function useActionBundles(statusFilter?: string | null, limit?: number) {
  return useQuery<AgentActionBundle[]>({
    queryKey: actionBundleKeys.list(statusFilter, null, limit),
    queryFn: async () => {
      return unwrap(api.listActionBundles(statusFilter ?? null, null, limit ?? null));
    },
  });
}

export function useSessionActionBundles(sessionId?: string | null, statusFilter?: string | null, limit?: number) {
  return useQuery<AgentActionBundle[]>({
    queryKey: actionBundleKeys.list(statusFilter, sessionId, limit),
    queryFn: async () => {
      return unwrap(api.listActionBundles(statusFilter ?? null, sessionId ?? null, limit ?? null));
    },
  });
}

export function useActionBundle(id: string | null) {
  return useQuery<AgentActionBundle | null>({
    queryKey: ["action-bundle", id],
    queryFn: async () => {
      if (!id) return null;
      return unwrap(api.getActionBundle(id));
    },
    enabled: id !== null,
  });
}

export function useApproveActionItem() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async (itemId: string) => {
      await unwrap(api.approveActionItem(itemId));
    },
    onSuccess: () => {
      invalidateDomains(qc, "agentActions");
    },
  });
}

export function useRejectActionItem() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async (itemId: string) => {
      await unwrap(api.rejectActionItem(itemId));
    },
    onSuccess: () => {
      invalidateDomains(qc, "agentActions");
    },
  });
}

export function useExecutionLog(bundleId: string | null) {
  return useQuery<AgentExecutionEntry[]>({
    queryKey: ["execution-log", bundleId],
    queryFn: async () => {
      if (!bundleId) return [];
      return unwrap(api.listExecutionLog(bundleId));
    },
    enabled: bundleId !== null,
  });
}

export function useExecuteActionBundle() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async (bundleId: string) => {
      return unwrap(api.executeActionBundle(bundleId));
    },
    onSuccess: () => {
      // Applying a bundle mutates the ledger (agentApply = agentActions +
      // transactions fan-out) and may fund goals; plus agent memory.
      void invalidateDomains(qc, "agentApply", "goals");
      void qc.invalidateQueries({ queryKey: ["agent-memory"] });
    },
  });
}
